File: src/main/java/catwalks/util/nestedmap/HierarchyKey.java
package catwalks.util.nestedmap;

import java.util.Arrays;

/**
 * Created by TheCodeWarrior
 */
class HierarchyKey {
	
	public static final ThreadLocal<MutableHierarchyKey> ACCESS_KEY = new ThreadLocal<MutableHierarchyKey>() {
		@Override
		protected MutableHierarchyKey initialValue() {
			return new MutableHierarchyKey();
		}
	};
	
	protected Object[] array;
	
	public HierarchyKey(Object[] array) {
		this.array = array;
	}
	
	
	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!( o instanceof HierarchyKey )) return false;
		
		HierarchyKey that = (HierarchyKey) o;
		
		// Probably incorrect - comparing Object[] arrays with Arrays.equals
		return Arrays.equals(array, that.array);
		
	}
	
	@Override
	public int hashCode() {
		return Arrays.hashCode(array);
	}
	
	public static class MutableHierarchyKey extends HierarchyKey {
		
		protected MutableHierarchyKey() {
			super(null);
		}
		
		public MutableHierarchyKey set(Object[] keys) {
			this.array = keys;
			return this;
		}
	}
}

File: src/main/java/catwalks/util/nestedmap/HierarchyKey.kt
package catwalks.util.nestedmap

import java.util.*

/**
 * Created by TheCodeWarrior
 */
internal open class HierarchyKey(protected var array: Array<out Any>?) {


    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is HierarchyKey) return false

// Probably incorrect - comparing Object[] arrays with Arrays.equals
        return Arrays.equals(array, o.array)

    }

    override fun hashCode(): Int {
        return Arrays.hashCode(array)
    }

    class MutableHierarchyKey : HierarchyKey(null) {

        fun set(keys: Array<out Any>): MutableHierarchyKey {
            this.array = keys
            return this
        }
    }

    companion object {

        val ACCESS_KEY: ThreadLocal<MutableHierarchyKey> = object : ThreadLocal<MutableHierarchyKey>() {
            override fun initialValue(): MutableHierarchyKey {
                return MutableHierarchyKey()
            }
        }
    }
}
